src/tasks/task-actuation-equality.cpp

```diff
@@ -15,11 +15,100 @@
 // <http://www.gnu.org/licenses/>.
 //
 
+#include <tsid/tasks/task-actuation-equality.hpp>
+#include "tsid/robots/robot-wrapper.hpp"
+
 namespace tsid
 {
   namespace tasks
   {
-    
-    
+    using namespace math;
+    using namespace pinocchio;
+
+    TaskActuationEquality::TaskActuationEquality(const std::string & name,
+                                             RobotWrapper & robot):
+     TaskActuation(name, robot),
+     m_constraint(name, robot.na(), robot.na())
+   {
+      m_ref = Vector::Zero(robot.na());
+      m_norm = Vector::Ones(robot.na());
+      Vector m = Vector::Ones(robot.na());
+      mask(m);
+    }
+
+    const Vector & TaskActuationEquality::mask() const
+    {
+      return m_mask;
+    }
+
+    void TaskActuationEquality::mask(const Vector & m)
+    {
+      assert(m.size()==m_robot.na());
+      m_mask = m;
+
+      const Vector::Index dim = static_cast<Vector::Index>(m.sum());
+      Matrix S = Matrix::Zero(dim, m_robot.na());
+      m_activeAxes.resize(dim);
+      unsigned int j=0;
+      for(unsigned int i=0; i<m.size(); i++)
+        if(m(i)!=0.0)
+        {
+          assert(m(i)==1.0);
+          S(j,i) = 1.0;
+          m_activeAxes(j) = i;
+          j++;
+        }
+      m_constraint.resize((unsigned int)dim, m_robot.na());
+      m_constraint.setMatrix(S);
+    }
+
+    int TaskActuationEquality::dim() const
+    {
+      return (int)m_mask.sum();
+    }
+
+    // Reference should be the same size as robot.na(), even if a mask is used (masked dof values will just be ignored)
+    void TaskActuationEquality::setReference(ConstRefVector ref)
+    {
+      assert(ref.size()==m_robot.na());
+      m_ref = ref;
+    }
+
+    const Vector & TaskActuationEquality::getReference() const
+    {
+      return m_ref;
+    }
+
+    // Normalization vector should be the same size as robot.na(), even if a mask is used (masked dof values will just be ignored)
+    void TaskActuationEquality::setNormalizationVector(ConstRefVector norm)
+    {
+      assert(norm.size()==m_robot.na());
+      m_norm = norm;
+
+      for(unsigned int i=0; i<m_activeAxes.size(); i++)
+        m_constraint.matrix()(i,m_activeAxes(i)) = m_norm(m_activeAxes(i));
+    }
+
+    const Vector & TaskActuationEquality::getNormalizationVector() const
+    {
+      return m_norm;
+    }
+
+    const ConstraintBase & TaskActuationEquality::getConstraint() const
+    {
+      return m_constraint;
+    }
+
+    const ConstraintBase & TaskActuationEquality::compute(const double ,
+                                                        ConstRefVector ,
+                                                        ConstRefVector ,
+                                                        Data & )
+    {
+      for(unsigned int i=0; i<m_activeAxes.size(); i++)
+        m_constraint.vector()(i) = m_ref(m_activeAxes(i)) * m_norm(m_activeAxes(i));
```

Comment:
Since the constraint vector depends only on `ref`, `norm` and `activeAxes`, wouldn't it make more sense to compute it only when one of these 3 quantities is updated?

Reply:
Agreed and done.